# Lambda School Data Science - A First Look at Data



In [3]:
## Lecture - let's explore Python DS libraries and examples!

#The Python Data Science ecosystem is huge. You've seen some of the big pieces - pandas, scikit-learn, matplotlib. What parts do you want to see more of?
"""
I'd like to see something completely different from what we're used to seeing i.e. Natural Language Processing with NLTK
"""

"\nI'd like to see something completely different from what we're used to seeing i.e. Natural Language Processing with NLTK\n"

# TODO - we'll be doing this live, taking requests
# and reproducing what it is to look up and learn things

Natural Language Tool Kit

## Assignment - now it's your turn

Pick at least one Python DS library, and using documentation/examples reproduce in this notebook something cool. It's OK if you don't fully understand it or get it 100% working, but do put in effort and look things up.

In [5]:
# TODO - your code here
# Use what we did live in lecture as an example

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.datasets import load_breast_cancer
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression

cancer = load_breast_cancer()

X_train, X_test, y_train, y_test = train_test_split(cancer.data, cancer.target, random_state=42)

scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

lr = LogisticRegression()
lr.fit(X_train, y_train)
lr.fit(X_train_scaled, y_train)

print("Test set score WITH scaled features: {:.2f}".format(lr.score(X_test_scaled, y_test)))
print("\nTest set score WITHOUT scaled features: {:.2f}".format(lr.score(X_test, y_test)))



Test set score WITH scaled features: 0.97

Test set score WITHOUT scaled features: 0.38


/Users/sammylee/miniconda3/envs/datascience/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
/Users/sammylee/miniconda3/envs/datascience/lib/python3.6/site-packages/sklearn/linear_model/logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [10]:
# Lots of help from Stack Overflow
df = pd.DataFrame(np.c_[cancer['data'], cancer['target']],
                  columns= np.append(cancer['feature_names'], ['target']))
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension   ...    worst texture  worst perimeter  worst area  \
0                 0.07871   ...            17.33           184.60      2019.0   
1                 0.05667   ...            23.41           158.80      1956.0   
2                 0.05999   ...            25.53           152.50      1709.0   
3                 0.09744   ...            26.50            98.87       567.7   
4                 0.05883   ...            16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890     0.0  
1          0.2750                  0.08902     0.0  
2          0.3613                  0.08758     0.0  
3          0.6638                  0.17300     0.0  
4          0.2364                  0.07678     0.0  

[5 rows x 31 columns]

In [12]:
df['target'].value_counts()

1.0    357
0.0    212
Name: target, dtype: int64

In [19]:
# After scaled Data using MinMaxScaler
df_2 = pd.DataFrame(X_test_scaled, columns=cancer['feature_names'])
df_2.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0     0.234047      0.300643        0.235951   0.133657         0.424483   
1     0.550908      0.392289        0.538341   0.411739         0.338178   
2     0.380479      0.330402        0.382558   0.248219         0.510698   
3     0.230619      0.269530        0.238654   0.127607         0.476393   
4     0.188501      0.159959        0.190141   0.099760         0.426198   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0          0.295939        0.187559             0.189911       0.404698   
1          0.286008        0.253046             0.395179       0.221570   
2          0.352442        0.343486             0.401938       0.407902   
3          0.384289        0.181373             0.139115       0.343833   
4          0.317170        0.157849             0.128926       0.347571   

   mean fractal dimension           ...             worst radius  \
0                0.290017           ...                 0.229954   
1                0.097936           ...                 0.591404   
2                0.168492           ...                 0.386741   
3                0.443555           ...                 0.153571   
4                0.376158           ...                 0.130911   

   worst texture  worst perimeter  worst area  worst smoothness  \
0       0.336354         0.211275    0.112718          0.485159   
1       0.388060         0.566367    0.407503          0.326903   
2       0.372601         0.357938    0.231342          0.566664   
3       0.290245         0.178537    0.072499          0.501460   
4       0.204158         0.123481    0.058108          0.430143   

   worst compactness  worst concavity  worst concave points  worst symmetry  \
0           0.231175         0.277998              0.348797        0.285630   
1           0.226562         0.279663              0.614777        0.194362   
2           0.232932         0.394567              0.520275        0.250739   
3           0.258739         0.250104              0.253265        0.195348   
4           0.202622         0.187032              0.237732        0.150601   

   worst fractal dimension  
0                 0.275178  
1                 0.091980  
2                 0.213208  
3                 0.326806  
4                 0.222957  

[5 rows x 30 columns]

### Assignment questions

After you've worked on some code, answer the following questions in this text block:

1.  Describe in a paragraph of text what you did and why, as if you were writing an email to somebody interested but nontechnical.

2.  What was the most challenging part of what you did?

3.  What was the most interesting thing you learned?

4.  What area would you like to explore with more time?




1. In the above example, I'm using Logistic Regression algorithm (to demonstrate the sklearn library) which makes predictions based on probability above a threshold (i.e. greater than 50% equals 1, less than 50% equals 0) to make predictions who would/would not get breast cancer based on characteristics such as mean radius and texture of the tumor tissue.  The above model has an accuracy score of 97%.

2. This actually wasn't challenging since the dataset automatically comes shipped with sklearn already pre-processed so the hardest task is already done.  

3. The most interesting thing I learned was just how much difference scaling can make as I show above.  This is the perfect example of "garbage in, garbage out" concept of statistical models.

4. I would love to explore one of the most well-tested innovations of data science/ML which is Natural Language Processing and its applications to businesses in the form of recommendation systems.

## Stretch goals and resources

Following are *optional* things for you to take a look at. Focus on the above assignment first, and make sure to commit and push your changes to GitHub (and since this is the first assignment of the sprint, open a PR as well).

- [pandas documentation](https://pandas.pydata.org/pandas-docs/stable/)
- [scikit-learn documentation](http://scikit-learn.org/stable/documentation.html)
- [matplotlib documentation](https://matplotlib.org/contents.html)
- [Awesome Data Science](https://github.com/bulutyazilim/awesome-datascience) - a list of many types of DS resources

Stretch goals:

- Find and read blogs, walkthroughs, and other examples of people working through cool things with data science - and share with your classmates!
- Write a blog post (Medium is a popular place to publish) introducing yourself as somebody learning data science, and talking about what you've learned already and what you're excited to learn more about